### PR TITLE
[SPARK-9570][Docs][YARN]Consistent recommendation for submitting spark apps to YARN, -master yarn --deploy-mode x vs -master yarn-x'

### PR DESCRIPTION
**Pull request by nssalian**
Issue link: https://issues.apache.org/jira/browse/SPARK-9570

Changes made:
1) Added the alternative to job submission to avoid the confusion
2) Moved the Master URLs section closer to the options prior to the examples

Requesting review.
Is there any other place in the documentation that could add a confusion to the user?
Need to maintain a consistent, if not clarify all the submission methods in the documentation.

### CODE REVIEW

File: docs/submitting-applications.md [hunk 1]
```diff
@@ -48,6 +48,44 @@ Some of the commonly used options are:
 * `application-jar`: Path to a bundled jar including your application and all dependencies. The URL must be globally visible inside of your cluster, for instance, an `hdfs://` path or a `file://` path that is present on all nodes.
 * `application-arguments`: Arguments passed to the main method of your main class, if any
 
+Alternatively, for submitting on yarn,
```
Comment (srowen) [inline docs/submitting-applications.md]: yarn -> YARN
I don't know if this helps as it's just presenting *both* ways to set this, and not clarifying which is preferred. It also duplicates the rest of the example.

File: docs/submitting-applications.md [hunk 2]
```diff
@@ -48,6 +48,44 @@ Some of the commonly used options are:
...
+./bin/spark-submit \
+  --class <main-class>
+  --master <yarn-deploy-mode>
+  --conf <key>=<value> \
+  ... # other options
+  <application-jar> \
+  [application-arguments] 
+{% endhighlight %}
+
+* `--master`: The --master parameter is either `yarn-client` or `yarn-cluster`. Defaults to `yarn-client`
```
Comment (tgravescs) [inline docs/submitting-applications.md]: actually I would rather see us use the --deploy-mode option to choose cluster vs client and --master yarn.

Unless something has changed we kept yarn-cluster and yarn-client around just for backwards compatibility.

Comment (nssalian) [inline docs/submitting-applications.md]: @srowen  and @tgravescs, users still have the confusion regarding the "recommended/preferred" method of submission. 
Not sure if it necessary to have a single method or have both ways.
I can modify the PR accordingly.

Comment (srowen) [inline docs/submitting-applications.md]: I don't think it makes sense to promote both ways. I also think both have to be retained for compatibility. I'm still not clear what people think the usual way to do it is; I've always seen and used `yarn-client` for example. In the end, after that's established, the docs should feature the preferred way, and mention the other in passing.

Comment (tgravescs) [inline docs/submitting-applications.md]: I agree we should have a preferred method and update docs to match.  I think the existing docs just never were fully updated. There wouldn't be a reason to have the --deploy-mode if we used yarn-client/yarn-cluster. There was discussion about it back then.

cc @sryza

Comment (nssalian) [inline docs/submitting-applications.md]: Will grep the docs to see the more popular approach to submission amongst the two.
Then align the docs to have that approach as a first recommendation and throw the latter as an alternative.
The goal is to have a consistent method overall.
Any suggestions?

**Comment by nssalian** [comment]
1) Running on YARN (which was fixed in #6924) has simply yarn-client and yarn-cluster for master.
and does not have `--deploy-mode` in the page.
2) For Standalone (https://spark.apache.org/docs/latest/spark-standalone.html), there is no such conflict in syntax just the explanation of the "deployment modes of Spark".
3) For Submitting applications (https://spark.apache.org/docs/latest/submitting-applications.html), there exist both the master `yarn-client` and `yarn-cluster` along with `--deploy-mode` since it is a holistic document for the submission and includes local, spark, mesos and yarn. But `--deploy-mode` is only used or appears in the examples to illustrate `supervise`, the rest just point to the master-urls.

@tgravescs, @srowen, the latest commits should help alleviate any confusion

**Comment by srowen** [comment]
I think the best question is, which is already used more in the docs? `--master yarn --deploy-mode client` or `--master yarn-client`? I count 3 uses of the former and 26 of the latter. I suggest therefore that:

- YARN docs continue to use `yarn-client` and `yarn-cluster`
- Add a mention about `--deploy-mode` after the first example of `yarn-client`/`yarn-cluster` in the YARN docs, that it's a synonym
- Don't rearrange the rest of the docs
- `spark-submit` help args should however use `--deploy-mode` since it's not YARN-specific
- Generic Spark submission docs should still use `--deploy-mode` for non-YARN modes

How does that strike people? It's much less change than eliminating `yarn-client`/`yarn-cluster` mentions, even if it highlights an inconsistency.

**Comment by tgravescs** [comment]
personally I would prefer us to change things to be consistent.  Change yarn docs to use --master yarn --deploy-mode client/cluster.   Yes it might be multiple instances of it but its only one or two document and should take little time.

**Comment by srowen** [comment]
Yeah I agree... though I also just now looked at the source code and many exception messages use `yarn-client` and `yarn-cluster`. It really seems that got embedded as the way to express this for YARN.

Any other feelings one way or the other? I'm pretty on the fence between consistency vs history.

**Comment by nssalian** [comment]
So, the consensus is to have `yarn-client` and `yarn-cluster` with the deploy-mode as alternative.
I agree with @srowen, more places in the code have `yarn-client` and `yarn-cluster`.

So the change is in the submitting applications doc wrt YARN and the Running on YARN doc wrt deploy-mode.
I'll change accordingly and update the PR soon.

**Comment by tgravescs** [comment]
sorry but I thought we were going with the --master yarn --deploy-mode client/cluster.

@sryza   I think you originally added the --deploy-mode was there discussion at that point about deprecating yarn-client/yarn-cluster?

**Comment by nssalian** [comment]
@tgravescs, I am not too sure whether to stick with consistency or history or both.
If @sryza  can weigh in, we can reach a good understanding of where to proceed on this.
The goal is to help a user and reduce the confusion (if any) in the submission methods for YARN.

**Comment by sryza** [comment]
Sorry all for the delay, have been in Indonesia for the week.

I'm also in favor of the standard way being --deploy-mode.  This keeps things consistent with standalone and Mesos.  I think it's probably fine if in some places in the docs we use "yarn-client" as shorthand to refer to running on YARN with the client deploy mode, but in the actual instructions on how to launch, my personal preference would be `--master yarn --deploy-mode client/cluster`.

**Comment by srowen** [comment]
@nssalian I would go with their suggestion. I suppose that means going all the way the other way, and featuring `--deploy-mode` in docs, comments and any log or exception messages, but then also mentioning `yarn-client` et al as a synonym. Both should remain supported and tested.

**Comment by nssalian** [comment]
@srowen, @sryza , @tgravescs  thank you for the feedback.
I will get this done this weekend.

**Comment by nssalian** [comment]
Creating a new PR.

**Pull request closed without merging**